v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,187 @@
+)
+
+type Cleaner struct {
+	apiExtensionsClient apiextensions.CustomResourceDefinitionInterface
+	client              client.Client
+	migrationBackoff    wait.Backoff
+	dryRun              bool
+	updated             int
+}
+
+func NewCleaner(apiExtensionsClient apiextensions.CustomResourceDefinitionInterface, client client.Client, dryRun bool) *Cleaner {
+	migrationBackoff := wait.Backoff{ // TODO: Still need to see if we want exponential backoff or linear is fine
+		Duration: 2 * time.Second, // wait 2s between attempts, this will help us in a state of conflict.
+		Steps:    3,               // 3 retry on error attempts per object
+	}
+	return &Cleaner{
+		apiExtensionsClient: apiExtensionsClient,
+		client:              client,
+		migrationBackoff:    migrationBackoff,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Cleaner) Run(ctx context.Context) error {
+	list, err := c.apiExtensionsClient.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if list == nil || len(list.Items) == 0 {
+		fmt.Println("found 0 results, make sure you have ASO CRDs installed")
+	}
+
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	deprecatedVersionRegexp := regexp.MustCompile(`v1alpha1api\d{8}(preview)?(storage)?`)
+	for _, crd := range list.Items {
+		crd := crd
+
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		newStoredVersions := removeMatchingStoredVersions(crd.Status.StoredVersions, deprecatedVersionRegexp)
+		objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, deprecatedVersionRegexp)
+		if err != nil {
+			return err
+		}
+
+		if len(newStoredVersions) > 0 && len(newStoredVersions) != len(crd.Status.StoredVersions) {
+
+			err = c.migrateObjects(ctx, objectsToMigrate)
+			if err != nil {
+				return err
+			}
+
+			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
+			if err != nil {
+				return err
+			}
+
+			c.updated++
+		} else {
+			fmt.Printf("nothing to update for '%s'\n", crd.Name)
+		}
+	}
+
+	if !c.dryRun {
+		fmt.Printf("updated %d CRD(s)\n", c.updated)
+
+	}
+
+	return nil
+}
+
+func (c *Cleaner) updateStorageVersions(
+	ctx context.Context,
+	crd apiextensionsapi.CustomResourceDefinition,
+	newStoredVersions []string) error {
+
+	if c.dryRun {
+		fmt.Printf("new storeVersions for '%s' CRD status storedVersions: %s\n", crd.Name, newStoredVersions)
+	} else {
+		crd.Status.StoredVersions = newStoredVersions
+		updatedCrd, err := c.apiExtensionsClient.UpdateStatus(ctx, &crd, v1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		fmt.Printf("updated '%s' CRD status storedVersions to : %s\n", crd.Name, updatedCrd.Status.StoredVersions)
+	}
+
+	return nil
+}
+
+func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstructured.UnstructuredList) error {
+	for _, obj := range objectsToMigrate.Items {
+		obj := obj
+		if c.dryRun {
+			fmt.Printf("resource '%s' to migrate for kind '%s'", obj.GetName(), obj.GroupVersionKind().Kind)
```

Comment:
minor: Both of these log messages might be quite verbose if there are a lot of resources. Do we need a `Logger` interface with verbosity defined, so that users can do `--v 2` or whatever in the cmdline to get more verbose logs, but by default they don't get these because they're kinda spammy?